### ISSUE TITLE
Bug: Can't download vagrant boxes from Jfrog artifactory with vagrant version 2.2.11 and up

### ISSUE THREAD

**Issue by shailygupta** [hashicorp/vagrant#12080]
### Vagrant version
2.2.11 and up

### Host operating system
macOS v10.15.7

### Vagrantfile

```ruby
require 'net/http'

@box_name_latest = 'testbox'
@box_url_local_jfrog = "https://<custom_jfrog_url>/artifactory/api/vagrant/vagrant/#{@box_name_latest}"
ENV['VAGRANT_SERVER_URL'] = "https://<custom_jfrog_url>/artifactory/api/vagrant"

Vagrant.configure('2') do |config|
  config.vm.box = @box_name_latest
  config.vm.box_url = @box_url_local_jfrog
end


```

### Debug output

https://gist.github.com/shailygupta/1e5c5b1286d3248ee840603f7b77ab6c

### Expected behavior

Box gets downloaded from artifactory using the ATLAS_TOKEN and VAGRANT_SERVER_URL. 
This worked up until vagrant version 2.2.10

### Actual behavior
```bash
An error occurred while downloading the remote file. The error
message, if any, is reproduced below. Please fix this error and try
again.

The requested URL returned error: 403 Forbidden

```

### Steps to reproduce

1. Install the latest vagrant version
2. Create a Vagrantfile similar to above
3. Set your ATLAS_TOKEN in your environment
4. Run `vagrant up` and try to downloade a box from artifactory (vagrant repo)

**Comment by CognexBela** [comment]
We are experiencing the exact same issue. After some investigation it seems that the problem is that the Authorization HTTP header introduced in 2.2.11 contains the token in % escaped encoding. This makes authentication fail on Artifactory. My understanding of RFC2616 & RFC2396 is that escaped encoding is only permitted for URI's, but not in HTTP headers so this seems to be a Vagrant issue.

**Comment by rockdreamer** [comment]
We are affected by this issue and have stopped upgrading vagrant because of this. This is also tracked in artifactory jira [here](https://www.jfrog.com/jira/browse/RTFACT-24061).

**Comment by rockdreamer** [comment]
[This looks like the related change](https://github.com/hashicorp/vagrant/commit/b5f6df9ec04b7a42e4a93961a38b667adbdcf06d#diff-2466e752be66035eb11fb4c7e7f690528de62adee331eecbef6e58cacd2d96a2R58).
@soapy1 I'm afraid my ruby-fu is not that good. Is is possible to change this code so that the token value doesn't get escaped? (@CognexBela does this sound right?)

**Comment by exit-code0** [comment]
I'm able to confirm this behavior with Vagrant 2.2.14 and Artifactory 7.4.1. No issues when trying the same with Vagrant 2.2.7.

**Comment by soapy1** [comment]
This issue should be resolved by https://github.com/hashicorp/vagrant/pull/12234 🎉

**Comment by rockdreamer** [comment]
Hi @soapy1 this is great news and thank you for the work on this issue. Is it possible to get a build for me to test?

**Comment by soapy1** [comment]
Heya, @rockdreamer one of these should work: https://github.com/hashicorp/vagrant-installers/releases/tag/v2.2.15.dev%2Bmain

**Comment by rockdreamer** [comment]
Thank you, I'll test right away :)

**Comment by rockdreamer** [comment]
I'm sorry to say that my issue has not been resolved. I downloaded the installer provided, ran the uninstall script and installed the new vagrant build.
I have set up my environment thus:
- `export ATLAS_TOKEN=$(curl -s -u$USER:$ARTIFACTORY_API_KEY "https://mydomain/artifactory/api/vagrant/auth")`
- export VAGRANT_DEFAULT_PROVIDER=vmware_fusion
- export VAGRANT_SERVER_URL="https://mydomain/artifactory/api/vagrant/my-vagrant-repo"

I went in a folder with an ubuntu box and ran vagrant box update --debug
The logs ended with: 
```
There was an error while downloading the metadata for this box.
The error message is shown below:

The requested URL returned error: 401
```

Results were similar with VAGRANT_CLOUD_TOKEN instead of ATLAS_TOKEN.
I'm attaching the output of the command. (Company name and dns name replaced)
[vagrant.log](https://github.com/hashicorp/vagrant/files/6200623/vagrant.log)

I'll be very glad to help out testing this and can make printf debuggging changes if pointed at the right place :)

**Comment by soapy1** [comment]
😢 oh no! that certainly looks like an issue. Will open this back up 👀 Thanks for taking the time to test this out!

**Comment by rockdreamer** [comment]
Hi @soapy1. I installed mitmproxy and investigated this further

What other alternatives I tried and possibilities of solving this:

`export RT_BOX=https://mydomain/artifactory/api/vagrant/my-vagrant-repo/boxname`
`curl -k $RT_BOX` returns
```
{
  "errors" : [ {
    "status" : 401,
    "message" : "Authentication is required"
  } ]
}
```

All reasonable

`curl -H "Accept: application/json" -H "Authorization: Bearer $ATLAS_TOKEN" -k $RT_BOX` returns
```
{
  "errors" : [ {
    "status" : 401,
    "message" : "Bad props auth token: basictoken=(contents of ATLAS_TOKEN)"
  } ]
}
```

This is the error we see with vagrant > 2.2.10

`curl -H "Accept: application/json" -k "$RT_BOX?access_token=$ATLAS_TOKEN"` returns the right result (this is the behaviour before 2.2.10).

I noticed that ATLAS_TOKEN is not urlencoded, so I tried saving the environment variable in urlencoded form. This had no effect.

Given we all use personal API keys for artifactory, I tried using its native form of authentication.

`curl -H "Accept: application/json" -H "X-JFrog-Art-Api:$ARTIFACTORY_API_KEY" -k $RT_BOX` returned the right results

I read a bit on what bearer tokens artifactory accepts. It's not pretty.
You have to [generate them using their API](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-CreateToken)

**Comment by rockdreamer** [comment]
TL;DR. It looks like artifactory really doesn't like the atlas token being passed as a Bearer token. It wants the token made in its own way.

My temporary workaround for this is to set ATLAS_TOKEN this way `export ATLAS_TOKEN=$(curl -s -u$USER:$ARTIFACTORY_API_KEY -XPOST "https://mydomain/artifactory/api/security/token" -d "username=USER" -d "scope=member-of-groups:*" | jq -r '.access_token')`

**Comment by chrisroberts** [comment]
@rockdreamer Thanks for providing this information. I just want to be sure I understand correctly. In the examples you provide `$ATLAS_TOKEN` and `$ARTIFACTORY_API_KEY` are two different values. Is that correct?

**Comment by chrisroberts** [comment]
Also, if you curl the `$RT_BOX` but include `?access_token=$ATLAS_TOKEN` on the end, do things work as expected?

**Comment by rockdreamer** [comment]
The UX of the workaround above is quite suboptimal but it gets the job done for a while. This token will only last for 3600 seconds. Non admins can't increase the expiration beyond that. Essentially there was usable integration between the products and now it feels like jumping through hoops.

@chrisroberts: 
- the ARTIFACTORY_API_KEY is read from the user's profile page and can be changed independently from the user's password.
- Artifactory provides an api endpoint (https://mydomain/artifactory/api/vagrant/auth) that will generate the ATLAS_TOKEN value. They document this [here](https://www.jfrog.com/confluence/display/JFROG/Vagrant+Repositories#VagrantRepositories-AuthenticatedAccesstoServers)
- Adding  `?access_token=$ATLAS_TOKEN ` at the end of the url makes the call succeed

Thoughts:
- Is it possible to go back to adding the access_token in the url?
- What would it take to change vagrant so that if ARTIFACTORY_API_KEY is present in the environment, it adds the `X-JFrog-Art-Api: $ARTIFACTORY_API_KEY` header to its curl calls? This seems like the easiest option to me. Is it something that can be sorted via a plugin?
- The artifactory docs mention a renewable bearer token but I have no idea how that might be used. Could that be an avenue?

Thank you both for looking into this :)

**Comment by chrisroberts** [comment]
@rockdreamer Having the access token work as expected is definitely the optimal solution. We have modified how we authenticate with Vagrant Cloud to reduce the visibility of the access token (since having it as a URL parameter makes it quite visible). Reading through the debug logs earlier, I had wrongly assumed that artifactory would work with the header which obviously is not the case :slightly_smiling_face: 

We did include an env var switch to allow the access token to be added to the URL parameters as an opt-in option. However, I'm not sure if artifactory will be upset that the header is set as well. Can you try setting the `VAGRANT_ALLOW_PARAM_AUTH_TOKEN` env var and seeing if it completes the request successfully?

Thanks so much!

**Comment by rockdreamer** [comment]
@chrisroberts I exported VAGRANT_ALLOW_PARAM_AUTH_TOKEN=1
vagrant adds the token in the url but also sends the Authentication: Bearer header.
Unfortunately artifactory rejects the token in the header and returns
```
{
    "errors": [
        {
            "message": "Bad props auth token: basictoken=non-urlencoded-contents-of-ATLAS_TOKEN)",
            "status": 401
        }
    ]
}
```

**Comment by chrisroberts** [comment]
Okay, I was afraid that would be the result. I think the best resolution for this will be to have the environment variable add the access token to the URL parameters _and_ disable the addition of the header auth. I'm also thinking that the env var we currently have defined will be renamed so it's clear that it will be a two part modification. 

Thanks for letting us know about the error you were still encountering and for running all these tests! We're just waiting on the openssl release to go live before we cut our next release tomorrow and I'm happy we can have this resolved before then.

Cheers!

**Comment by tomasnorre** [comment]
According to GitHub Releases the 2.2.15 is already release.
https://github.com/hashicorp/vagrant/releases

but on https://www.vagrantup.com/downloads it's still 2.2.14 listed as the latest version.

**Comment by tomasnorre** [comment]
If I go directly to https://releases.hashicorp.com/vagrant/2.2.15/ then I have the new releases. 
It's "just" the website not up to date. 

Unfortunately it didn't solve my problem completely, but think it's a local issue now.

**Comment by chrisroberts** [comment]
@tomasnorre Please note that you'll need to set the [VAGRANT_SERVER_ACCESS_TOKEN_BY_URL](https://www.vagrantup.com/docs/other/environmental-variables#vagrant_server_access_token_by_url) environment variable to get access token usage working with artifactory. If you are still encountering unexpected problems, please just open up a new issue. :slightly_smiling_face:

**Comment by tomasnorre** [comment]
Thanks, that I have overlooked. Will check it out tomorrow. My problem currently is regarding missing metadata.json, but will check you hint as well. 

Thanks.
